pghoard/basebackup/delta.py

```diff
@@ -0,0 +1,479 @@
+                        relative_path=snapshotfile.relative_path,
+                        hexdigest=new_hash,
+                        file_size=size,
+                        stored_file_size=stored_size
+                    )
+                    if not skip_upload:
+                        # Every thread is waiting for the transfer to finish, so we don't load disk too much,
+                        # only `self.parallel` files in parallel
+                        while True:
+                            try:
+                                callback_queue.get(timeout=queue_timeout)
+                                break
+                            except Empty:
+                                continue
+            except FileNotFoundError:
+                if snapshotfile.missing_ok:
+                    self.log.warning("File disappeared while uploading: %r", path)
+                    return True
+                else:
+                    self.log.error("Required file disappeared while uploading: %r, aborting", path)
+                    return False
+            except Exception:  # pylint: disable=broad-except
+                # Report failure - whole step will be retried later
+                self.log.exception("Exception uploading %r", path)
+                return False
+
+        return True
+
+    def _upload_single_delta_files(
+        self, todo_hexdigests: Set[str], snapshotter: Snapshotter, progress: float
+    ) -> UploadedFilesMetric:
+        """Upload delta files, if upload fails - try to clean up uploaded files"""
+        # Keep track of the new hashes submitted with the current upload, so we can clean them up in case of error
+        new_hashes: Dict[str, Tuple[int, int]] = {}
+        callback_queue = CallbackQueue()
+        hexdigests_count = len(todo_hexdigests)
+
+        sorted_todo_hexdigests = sorted(
+            todo_hexdigests, key=lambda hexdigest: -snapshotter.hexdigest_to_snapshotfiles[hexdigest][0].file_size
+        )
+        iterable_as_list = list(sorted_todo_hexdigests)
+        submit_files = partial(
+            self._submit_files_in_thread, snapshotter=snapshotter, callback_queue=callback_queue, new_hashes=new_hashes
+        )
+        with Pool(self.parallel) as p:
+            for idx, hexdigest_result in enumerate(zip(iterable_as_list, p.imap(submit_files, iterable_as_list))):
+                hexdigest, res = hexdigest_result
+                if not res:
+                    self.log.error(
+                        "Error while processing digest for upload %r, waiting for workers pool to shutdown", hexdigest
+                    )
+
+                    p.terminate()
+                    p.join()
+
+                    self.log.info("Cleaning up already uploaded new backup files")
+
+                    for new_hash in new_hashes:
+                        key = os.path.join(self.site_config["prefix"], "basebackup_delta", new_hash)
```

Comment:
Same here, we should avoid those "magic strings" and use the FileTypePrefixes dict.